File: Source code/api/src/models/Product.ts
import { Schema, model, Types } from 'mongoose';

export interface ILiveCamera {
  _id?: Types.ObjectId;
  name: string;
  stream_url: string;
  location?: string;
  is_active: boolean;
  thumbnail?: string;
}

export interface IProduct {
  name: string;
  category: string;
  type: 'Plant' | 'Animal';
  description: string;
  origin: string;
  cultivation_time?: string;
  initial_quantity: number;
  current_quantity: number;
  unit: string;
  images: { path: string; filename: string }[];
  live_cameras: ILiveCamera[];
  farming_area?: Types.ObjectId;
  parent_product?: Types.ObjectId;
  source_products?: Types.ObjectId[];
  qrcode?: string;
  status: 'draft' | 'active' | 'completed' | 'recalled';
  onChainBatchId?: string;
  created_by: Types.ObjectId;
  isDeleted: boolean;
  deletedAt?: Date;
  deleted_by?: Types.ObjectId;
}

const productSchema = new Schema<IProduct>(
  {
    name: {
      type: String,
      required: [true, 'Vui l?ng nh?p t?n s?n ph?m'],
      trim: true,
      maxlength: 200,
    },
    category: {
      type: String,
      required: [true, 'Vui l?ng nh?p danh m?c'],
      trim: true,
    },
    type: {
      type: String,
      required: [true, 'Vui l?ng ch?n lo?i s?n ph?m'],
      enum: {
        values: ['Plant', 'Animal'],
        message: '{VALUE} kh?ng h?p l?',
      },
    },
    description: {
      type: String,
      required: [true, 'Vui l?ng nh?p m? t?'],
    },
    origin: {
      type: String,
      default: 'Vi?t Nam',
    },
    cultivation_time: {
      type: String,
    },
    initial_quantity: {
      type: Number,
      min: [0, 'Số lượng ban đầu không được âm'],
      default: 0,
    },
    current_quantity: {
      type: Number,
      min: [0, 'Số lượng tồn không được âm'],
      default: 0,
      index: true,
    },
    unit: {
      type: String,
      trim: true,
      default: 'kg',
      maxlength: 30,
    },
    images: {
      type: [{ path: String, filename: String }],
      default: [],
    },
    live_cameras: {
      type: [
        {
          name: {
            type: String,
            required: [true, 'Vui lòng nhập tên camera'],
            trim: true,
            maxlength: 120,
          },
          stream_url: {
            type: String,
            required: [true, 'Vui lòng nhập URL live stream'],
            trim: true,
          },
          location: {
            type: String,
            trim: true,
            maxlength: 200,
          },
          is_active: {
            type: Boolean,
            default: true,
          },
          thumbnail: {
            type: String,
            trim: true,
          },
        },
      ],
      default: [],
    },
    farming_area: {
      type: Schema.Types.ObjectId,
      ref: 'FarmingArea',
    },
    parent_product: {
      type: Schema.Types.ObjectId,
      ref: 'Product',
    },
    source_products: [
      {
        type: Schema.Types.ObjectId,
        ref: 'Product',
      },
    ],
    qrcode: {
      type: String,
    },
    status: {
      type: String,
      enum: ['draft', 'active', 'completed', 'recalled'],
      default: 'draft',
    },
    onChainBatchId: {
      type: String,
    },
    created_by: {
      type: Schema.Types.ObjectId,
      ref: 'User',
      required: true,
    },
    isDeleted: {
      type: Boolean,
      default: false,
      index: true,
    },
    deletedAt: {
      type: Date,
    },
    deleted_by: {
      type: Schema.Types.ObjectId,
      ref: 'User',
    },
  },
  { timestamps: true }
);

export default model<IProduct>('Product', productSchema);

File: Source code/app/lib/models/product.dart
import 'package:app/models/live_camera.dart';

class Product {
  final String id;
  final String name;
  final String? description;
  final String category;
  final String origin;
  final String type; // plant | animal
  final double initialQuantity;
  final double currentQuantity;
  final String unit;
  final List<String> images;
  final List<LiveCamera> liveCameras;
  final String? onChainBatchId;
  final String status; // draft | active | completed | recalled
  final Map<String, dynamic>? farmingArea;
  final DateTime? cultivationTime;
  final Map<String, dynamic>? createdBy;
  final DateTime? createdAt;
  final bool isDeleted;
  final DateTime? deletedAt;
  final Map<String, dynamic>? deletedBy;

  Product({
    required this.id,
    required this.name,
    this.description,
    required this.category,
    required this.origin,
    required this.type,
    this.initialQuantity = 0,
    this.currentQuantity = 0,
    this.unit = 'kg',
    required this.images,
    this.liveCameras = const [],
    this.onChainBatchId,
    required this.status,
    this.farmingArea,
    this.cultivationTime,
    this.createdBy,
    this.createdAt,
    this.isDeleted = false,
    this.deletedAt,
    this.deletedBy,
  });

  /// `batchId` dùng trên blockchain đang lấy trực tiếp từ `_id`.
  String get batchId => id;

  /// Nhãn hiển thị loại sản phẩm
  String get typeLabel => type == 'animal' ? 'Chăn nuôi' : 'Trồng trọt';
  String get typeIcon => type == 'animal' ? '🐄' : '🌾';

  /// Nhãn hiển thị trạng thái
  String get statusLabel {
    switch (status) {
      case 'active': return 'Đang sản xuất';
      case 'completed': return 'Hoàn thành';
      case 'recalled': return 'Đã thu hồi';
      default: return 'Nháp';
    }
  }

  /// Tên người tạo
  String get createdByName {
    if (createdBy == null) return 'N/A';
    final first = createdBy!['first_name'] ?? '';
    final last = createdBy!['last_name'] ?? '';
    return '$first $last'.trim();
  }

  factory Product.fromJson(Map<String, dynamic> json) => Product(
    id: json['_id'] as String,
    name: json['name'] as String,
    description: json['description'] as String?,
    category: json['category'] as String? ?? '',
    origin: json['origin'] as String? ?? '',
    type: json['type'] as String? ?? 'plant',
    initialQuantity: _numToDouble(json['initial_quantity']),
    currentQuantity: _numToDouble(json['current_quantity']),
    unit: json['unit']?.toString() ?? 'kg',
    images: _parseImages(json['images']),
    liveCameras: (json['live_cameras'] as List<dynamic>? ?? const [])
        .whereType<Map<String, dynamic>>()
        .map(LiveCamera.fromJson)
        .toList(),
    onChainBatchId: json['onChainBatchId'] as String?,
    status: json['status'] as String? ?? 'draft',
    farmingArea: json['farming_area'] as Map<String, dynamic>?,
    cultivationTime: json['cultivation_time'] != null
        ? DateTime.tryParse(json['cultivation_time'] as String)
        : null,
    createdBy: json['created_by'] as Map<String, dynamic>?,
    createdAt: json['createdAt'] != null
        ? DateTime.tryParse(json['createdAt'] as String)
        : null,
    isDeleted: json['isDeleted'] == true,
    deletedAt: json['deletedAt'] != null
        ? DateTime.tryParse(json['deletedAt'] as String)
        : null,
    deletedBy: json['deleted_by'] as Map<String, dynamic>?,
  );

  static List<String> _parseImages(dynamic raw) {
    if (raw == null) return [];
    if (raw is! List) return [];
    return raw
        .map((e) {
          if (e is Map) return (e['path'] as String?) ?? '';
          return e.toString();
        })
        .where((s) => s.isNotEmpty)
        .toList();
  }

  static double _numToDouble(dynamic raw) {
    if (raw is num) return raw.toDouble();
    return double.tryParse(raw?.toString() ?? '') ?? 0;
  }
}
